FEAT ADD MISSING METHODS FOR INTEGRATIONS #70

fadkeabhi opened the pull request:
added method setPrimary for endpoint /v1/integrations/{integrationID}/set-primary 
### Used For
Set an integration as primary
### Args:
integration_id: The integration ID

### Returns:
Bool : Is integration set active

### Testing : 
I have tested the changes on my local machine. The function is returning True if the integration with given Id is set to Primary. 
HTTPError: 400 is raised if the integration id is incorrect from the handle_request method. As there is no way to get response for handle_request method if the responce from API is other than 200.

unicodeveloper commented:
@ryshu Please can you help review this PR?

mr-palindrome commented:
Hey @fadkeabhi it seems you have missed writing the test for the method that you have added! Can you do that as well!

ryshu commented on novu/api/integration.py:
```diff
@@ -109,3 +109,21 @@ def limit(self, channel: Channel) -> IntegrationChannelUsageDto:
+
+
+
+    def setPrimary(self, integration_id: str) -> bool:
```
All methods should be declared using "snake_case"

```suggestion
    def set_primary(self, integration_id: str) -> bool:
```

ryshu commented on novu/api/integration.py:
```diff
@@ -109,3 +109,21 @@ def limit(self, channel: Channel) -> IntegrationChannelUsageDto:
+        """
+
+        res = self.handle_request("POST", f"{self._integration_url}/{integration_id}/set-primary")["data"]
+        if res.get("active")==True:
+            return True
+        else:
+            return False
```
Could be inline, which seen better to me

```suggestion
        return res.get("active") == True
```

ryshu reviewed (changes requested):
Missing unit tests

Also, you need to install pre-commit locally which will help you to pre-check your code is properly formatted and your commit match convention

See https://github.com/novuhq/novu-python/blob/main/CONTRIBUTING.rst for more details about how to contribute 😉

fadkeabhi commented:
Now I have added unit test for the function. Also added a try catch statement in function, as when function is called with an invalid integration_id, the endpoint responded with 404 error. Which is not handled anywhere. leading to crashing of application.

fadkeabhi commented:
@ryshu @mr-palindrome Can you take a look at it again. Additionally I tested the updated SDK with my own credentials. The functionality is as intended.

ryshu commented on novu/api/integration.py:
```diff
@@ -109,3 +109,22 @@ def limit(self, channel: Channel) -> IntegrationChannelUsageDto:
+        try:
+            res = self.handle_request("POST", f"{self._integration_url}/{integration_id}/set-primary")["data"]
+            return res.get("primary") == True
+        except:
```
Broad-except aren't allowed, and i don't understand what you are trying to catch here ?

fadkeabhi replied:
Now I have changed Broad-except to only accept the http errors. As when the endpoint set-primary is called with an invalid integration_id it responds with the 400 error. And in handle_request module of api/base.py the 400 error is not handled. Which is further leading to an crash of entire SDK and eventually entire application will crash.

ryshu commented on tests/api/test_integration.py:
```diff
@@ -411,3 +412,19 @@ def test_limit(self, mock_request: mock.MagicMock) -> None:
+
+
+    @mock.patch("requests.request")
+    def test_set_primary(self, mock_request: mock.MagicMock) -> None:
```
You might need to add a test for 404 behavior

ryshu reviewed (changes requested):
Pylama and precommit tests fail.

You should read and apply the instructions in the contribution guidelines if you don't want there to be round trips on your MR: https://github.com/novuhq/novu-python/blob/main/CONTRIBUTING.rst

For example, to install the project you are required to install the precommit which ensures that the code formatting respects the configuration.

ryshu commented on novu/api/integration.py:
```diff
@@ -109,3 +109,18 @@ def limit(self, channel: Channel) -> IntegrationChannelUsageDto:
             self.handle_request("GET", f"{self._integration_url}/{channel}/limit")["data"]
         )
+
+    def set_primary(self, integration_id: str) -> bool:
+        """Set an integration as primary
+
+        Args:
+            integration_id: The integration ID
+
+        Returns:
+            Is integration set active
+        """
+        try:
+            res = self.handle_request("POST", f"{self._integration_url}/{integration_id}/set-primary")["data"]
+            return res.get("primary") is True
+        except requests.exceptions.HTTPError:
+            return False
```
I still disagree with your proposal. I don't think it's actually a good idea to make the decision for the user how to handle errors.

If we want to warn our users that this could go wrong, then we might as well just specify it in the documentation. They will be able to define the procedure to follow in this case themselves.

We are a connector, not a decision-making tool.

```suggestion
    def set_primary(self, integration_id: str) -> IntegrationDto:
        """Set an integration as primary

        Args:
            integration_id: The integration ID

        Raises:
            HTTPError: Raise a 400 error when the integration identifier is poorly formulated
            HTTPError: Raise a 403 error when you are not authorized to do this action
            HTTPError: Raise a 404 error when provided integration identifier is not found in the application

        Returns:
            The instance of the updated integration
        """
        return IntegrationDto.from_camel_case(
            self.handle_request("POST", f"{self._integration_url}/{integration_id}/set-primary")["data"]
        )
```

Is you want to provide an utils

fadkeabhi replied:
I have made changes as per your suggestions. Also both testcases are modified accordingly.

ryshu commented:
@unicodeveloper Looking at the discussion on the outcome, perhaps we should close this MR?

unicodeveloper commented:
If we close it, how would we get this method/PR in? @ryshu

ryshu commented:
> If we close it, how would we get this method/PR in? @ryshu

This is an error, I had not seen that the private method of the API concerned the other method which is not implemented here
